### ISSUE TITLE
Weird compiler behavior with in-lining of a function -> constant value arbitrary changed

### ISSUE THREAD

**Issue by 2ndTaleStudio** [rust-lang/rust#63516]
Hey there,

I guess I found a weird bug in the compiler in the following scenario:
I do have a function with the following signature and the first part of the contents given like this:
```
pub(crate) fn send_message<T: MailboxMessage>(channel: MailboxChannel, message: &T) -> MailboxResult<&T> {
    let msg_ptr: *const T = message;
    let msg_ptr_uncached: u32 = (msg_ptr as u32) | 0xC000_0000;
```
When this function get's compiled into a binary and the compiler decides to inline it, the constant value for the OR operation is changed to 0xC000_000**8**. The assembly looks like this:
```
[...]
8e08:	e5901000 	ldr	r1, [r0]
    8e0c:	e3510000 	cmp	r1, #0
    8e10:	4afffffc 	bmi	8e08 <_ZN15ruspiro_mailbox7Mailbox13get_clockrate17hf8c4e9c932479a4dE+0x48>
    8e14:	e59f1098 	ldr	r1, [pc, #152]	; 8eb4 <_ZN15ruspiro_mailbox7Mailbox13get_clockrate17hf8c4e9c932479a4dE+0xf4>
    8e18:	e3850123 	orr	r0, r5, #-1073741816	; 0xc0000008
[...]
```
This lead to very subtile error as the resulting value is just wrong!
Interestingly when forcing the compiler never to inline the function with ``#[inline(never)]``
The resulting assembly is correct when it comes to the OR operation:
```
8e7c:	e5901000 	ldr	r1, [r0]
    8e80:	e3510000 	cmp	r1, #0
    8e84:	4afffffc 	bmi	8e7c <_ZN15ruspiro_mailbox9interface12send_message17he410abe85cc9a455E+0x1c>
    8e88:	e59f207c 	ldr	r2, [pc, #124]	; 8f0c <_ZN15ruspiro_mailbox9interface12send_message17he410abe85cc9a455E+0xac>
    8e8c:	e6ef0076 	uxtb	r0, r6
    8e90:	e1801005 	orr	r1, r0, r5
    8e94:	e3811103 	orr	r1, r1, #-1073741824	; 0xc0000000
```
Forcing the compiler not to inline the function is a working solution in the scenario, but as it was quite hard to find this bug I'm wondering whether this subtle issue exists on other places in the binary as well which lead to bugs I've not yet seen in my application.

I'm using rust-nightly and cross compiling from windows host machine to armv7-unknown-linux-gnueabihf
```
rustc 1.38.0-nightly (2d1a551e1 2019-08-08)
rustup 1.18.3 (435397f48 2019-05-22)
cargo 1.38.0-nightly (42a8c0adf 2019-08-07)
xargo 0.3.13
```

**Comment by jonas-schievink** [comment]
Can you provide the full source code that reproduces this behavior?

**Comment by 2ndTaleStudio** [comment]
Hi,
sure. The whole source code of a reproducable example could be found here:
https://github.com/RusPiRo/ruspiro-loader/tree/master/compile-issue

If you patch the ``ruspiro-mailbox`` crate and change the file ``interface.rs`` there from:
```
#[inline(never)] // never inline, if inlined the compiler seem to mess up the | 0xC000_0000 and do a | 0xC000_0008?????
pub(crate) fn send_message<T: MailboxMessage>(channel: MailboxChannel, message: &T) -> MailboxResult<&T> {
```

to:
```
#[inline(always)] // never inline, if inlined the compiler seem to mess up the | 0xC000_0000 and do a | 0xC000_0008?????
pub(crate) fn send_message<T: MailboxMessage>(channel: MailboxChannel, message: &T) -> MailboxResult<&T> {
```

you can see the different binaries being created. I'm using the make file in this folder to build, but it depents on some path set on my machine and might need adoption at your end...
Thanks in advance for taking a look.

**Comment by Centril** [comment]
Pending investigation I'm labeling this as a soundness hole under the idea of "wrong dynamic semantics + unsafe code can assume right semantics ==> boom!". Also labeling as P-high for now.

**Comment by nikomatsakis** [comment]
Compiler team update:

This looks pretty clearly like a problem caused by LLVM's optimizations, though it could be rustc's fault for generating invalid data. The source code in [this comment](https://github.com/rust-lang/rust/issues/63516#issuecomment-520834166) looks *fairly* minimal. It really needs somebody to try and determine if this is an LLVM bug or what.

**Comment by rkruppe** [comment]
The reproduction in https://github.com/RusPiRo/ruspiro-loader/tree/master/compile-issue is small at first look but pulls in a number of non-trivial crates. It would help if someone could minimize it to a smaller single-crate (or perhaps two-crate since inlining is relevant) program that can be built just with a handful of compiler invocations. Ideally, the linker script and other non-standard build system shenanigans could also be stripped away to make it easier to reproduce without having a C cross-toolchain set up.

**Comment by nikomatsakis** [comment]
Ah, ok, I missed the dependencies =)

**Comment by ollie27** [comment]
I think `0xc0000008` is from inlining `(data & 0xFFFF_FFF0) | ((channel as u8) & 0xF) as u32` from https://github.com/RusPiRo/ruspiro-mailbox/blob/baf443680845ef6331bda9886010b83f9d0aedf8/src/interface.rs#L107. You can see the same constant in the assembly for similar code on the [playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=674002775697bd8e8daf24510e804fb8):

```
	orl	$-1073741816, %edi      # imm = 0xC0000008
```

That looks like a perfectly legal optimisation so if the code isn't working I reckon something else is wrong.

**Comment by 2ndTaleStudio** [comment]
@ollie27 Well nice spotted, but I think this is not a generic valid optimization...
Check the following please:
```
let addr:u32 = 0x1A0;
    let addr_uc:u32 = addr | 0xC000_0000; // will be 0xC000_01A0
    let channel:u32 = 3;
    let data:u32 = (addr_uc & 0xFFFF_FFF0) | (channel & 0xF); // will be 0xC000_01A3
    let data_opt:u32 = addr_uc | 0xC000_0008; // will be 0xC000_01A8
    
    println!("data:{:X}, opt:{:X}", data, data_opt);
```
The output is: ``data:C00001A3, opt:C00001A8``
So the optimization is only valid in case the channel value is 8.
However, even if this optimisation is fine and only be the visual spotted symptom I'm still struggling with the fact, that inlining a function might cause different behavior. I tried hard to produce a smaller example (without much luck as I'm only able to test the functionality on real Raspberry Pi). 
 
I'm also trying to figure out if the root cause of the issue is somewhere else and not a compiler issue as the failing behavior is rather hard to reproduce...

I've changed the beginning of the ``send_message`` function in the ``ruspiro-mailbox`` crate to print stuff to the console. Based on this I found the reproduce-able setup:
```
pub(crate) fn send_message<T: MailboxMessage>(channel: MailboxChannel, message: &T) -> MailboxResult<&T> {
    let msg_ptr: *const T = message;
    let msg_ptr_uncached: u32 = msg_ptr as u32 | 0xC000_0000;
    print("simple print\r\n"); // works
    print!("{}", "print in macro print!\r\n"); //works
    print(format!("{}\r\n", "simple print with format!().as_str()").as_str()); // works
    print(format!("{}", format!("{}\r\n", "print with format!(format!().as_str()).as_str()").as_str()).as_str()); // fails
```
The console output is like this:
```
simple print
print in macro print!
simple print with format!().as_str()
print with format!(format!().as_str()).as_str()
```
Interestingly the function does what it should with those print statement only if it is **not** inlined. As soon as the last one - format! within in format! macro is used the ``send_message`` function fails when inlined.

For reference: The ``format!`` macro is the one from the ``alloc crate``.
The ``print`` function puts a string to the Uart console. The signature is:
``pub fn print(s: &str)``
the ``print!`` macro looks like this:
```
macro_rules! print {
    //$crate::macros::alloc::
    ($($arg:tt)*) => ($crate::print($crate::alloc::format!($($arg)*).as_str()));
}
```

May be this all is just because of my lack of knowledge or experience in Rust but what drives me crazy a bit is, why it makes a difference if the compiler ``inline`` the function or not...

Thanks for taking your time to read and investigate into this...

**Comment by gnprice** [comment]
> So the optimization is only valid in case the channel value is 8.

From looking a bit at the code you linked to, it appears that this is in fact the case in the only calls to your `send_message` function ([here, I think](https://github.com/RusPiRo/ruspiro-mailbox/blob/baf443680845ef6331bda9886010b83f9d0aedf8/src/lib.rs) -- where 8 is spelled `MailboxChannel::PropertyTagsVc`) in any of the Rust source code in the project.

In general, the compiler has the freedom to move code in either direction between a function and its callers, when it knows it's seeing every possible caller there will be. That's an important freedom for many optimizations. So it's perfectly normal to look at a function's machine code and find that it doesn't match the semantics of its source code -- it's only the program as a whole that has to behave the way the source code says.

It sounds like you didn't start by looking at the machine code, though -- it sounds like you observed some buggy behavior in your program first. You haven't said how you're producing actual behavior from this code that isn't what you expected... but given your [linker script](https://github.com/RusPiRo/ruspiro-loader/blob/master/compile-issue/link.ld) and your mention of a bootloader, I'm guessing that perhaps there's code somewhere else involved, not in Rust (or in Rust but built separately and then manually linked), which is looking for the symbols in this binary and expecting to be able to call the code corresponding to `send_message` and have it behave like the source code of `send_message` says.

In that case, you'll need to inform the compiler that there are other non-Rust callers of this function out there, so it knows it can't optimize across that function boundary at will. This is expressed with the `extern` keyword:
https://doc.rust-lang.org/std/keyword.extern.html
https://doc.rust-lang.org/reference/items/functions.html#extern-functions
You might find `#[no_mangle]` convenient too.

**Comment by 2ndTaleStudio** [comment]
Hi, thanks again for your time looking into this and for the clarification on the "freedoms" of the compiler ;). You are right that I've missed to provide some further context and you are also right that looking at the compiler output was the attempt to understand why a certain peace of my program does not work.

Please allow me to give some more background, that will hopefully also explain why the ``extern`` or ``#[no_mangle]`` options does not suite here from my point of view.

## Overall Context
I'm building a kernel for a Raspberry Pi 3 written as much as possible in Rust. The first part of the kernel is written in assebly to boot up the cores and prepare the environment to be able to branch into the functions written in Rust. Once the first "main" function is called inside Rust there is - appart from interrupts - nothing that is calling from the outside into the Rust environment.

## Context around the issue
The raspberry Pi is running an ARM CPU and a GPU. The GPU is kicked off with it's own firmware delivered by the vendor. This firmware allows the CPU to request specific data from the GPU using so called Property Tag messages. To do so, one creates the message on either the HEAP or the STACK and passes the address of this message using a specific memory mapped hardware register. This address need to be aligned to 16 bit and the content this address points to have a specific format, that is out of scope for this issue I guess. When passing the address into the mentioned register it need to be "aliased" with 0xC000_0000 to ensure the GPU can access the contents while by-passing any cache. The LSByte is used to pass the message channel for this message. 

## The issue itself
I've written a function that prepares such a property tag message with right memory layout etc. on the stack.
I've written another generic function that is able to take any of such messages and pass them to the GPU with the procedure described above. It takes the address of the message provided, alias with 0xC000_0000, add's the channel number and writes the same into the corresponding register. This function than checks the result of this transfer and waits on another MMIO register until the GPU has signaled the message has been processed and the response is available. The function than returns the response from the GPU that writes the response to the exact same memory location given when sending the message. (I might need to mention that the Message format defines a ``union`` for the "payload" that will contain the request data when sending and the response data after receiving, it is ensured that the whole message size is always the max size of either request or response).

But here comes now the behavior I do not fully get:
Lets name the specific function for a particular message:
``fn send_specific_msg()`` and the one for the generic part ``fn send_generic_msg<T>(msg: &T)``.
Without changing the contents of both function, but controlling whether the generic function gets inlined as a compiler optimization or not seem to influence the whole behavior, that the message passed to the GPU could be processed or not. As I've written the same functionality already in C/C++ to proof that at least the approach is working and the GPU is behaving in reproducible manner I did assume the Rust compiler might mess up things - either on the stack or with the address passed into the register for the GPU to process. Here I spotted the difference with the ``OR 0xC000_0008`` instead of ``0xC000_0000`` in case of inlining which as you clearly pointed out is definitely not the root cause.

This is where I'm stuck now... with my lack of knowledge on the Rust compiler and the very beginner knowledge of Rust at all I'm trying to do various variants of the functions with inline without inline with some console statements inside the functions to see what might make the difference with the different call variants and I'm not able to spot it. It might be trivial and it might not be a compiler issue and only happening in my special case due to how I'm using crate dependency and the ``no_std`` environment, but it might also be an indication of some very subtle issue that you or others much more experience than I could find....

Thx in advance for being patient with me and helping in this topic.

**Comment by gnprice** [comment]
Got it. OK, if you're only ever calling your `send_message` from those callsites where the first argument is indeed 8, then the 0xc000_0008 is just a correct optimization -- not only in terms of the language semantics that the compiler promises to respect, but also in terms of the ways you're trying to use the result.

At this point it sounds like you have a debugging challenge. :wink: The best place for further detailed discussion on it won't be this thread; a post on https://users.rust-lang.org/ might be a good place to get help.

Just one quick observation here: it sounds like you're sending to the GPU the address of a struct you've placed in memory. You're promising the GPU the struct will be there when it goes and reads at that address... but Rust doesn't know about that promise. I suspect that in your code, once you send that message, the struct's lifetime has ended, and Rust considers that struct's memory free to be re-used for something else. To make things work correctly, you'll want to ensure that the struct's lifetime extends past when the GPU is done reading it (e.g., maybe by ensuring it extends past when you read the GPU's response.)

There's lots to say about lifetimes in Rust, so if that paragraph doesn't entirely make sense, I recommend consulting a source like the Rust book and/or asking a question on a forum like users.rust-lang.org.

Comparing to C or C++: this type of bug corresponds exactly to calling `free` or `delete` too soon. The magic of lifetimes is that that class of bug (among others) becomes impossible... but only as long as the pointers stay within safe Rust, and don't get dereferenced by either unsafe Rust or code outside Rust.

**Comment by 2ndTaleStudio** [comment]
Hi,
well true it is a challenge and thanks for bringing up the lifetime topic and I would love to see that this would be the solution to the issue. But ( sorry that there is a but ;) )

The function that is sending the data address to the GPU is also waiting for the GPU to acknowledge that it has finished processing and passed the response to that address. So from my point of view the promise I'm giving from rust point of view is adhered. However, to further allow rust to know about this I added the lifetime specifier to my ``send_generic_msg`` function so it's even more ensured the lifetime of the struct's reference is ensured like so:

```
fn send_specific_msg() {
    send_specific_msg(0x8, &MyMessage::new());
}

fn send_generic_msg<'a, T: Message>(channel: u8, msg: &'a T) {
   write_gpu().and_then(|| {
     read_gpu().and_ten(|| { })
  });
}
```

And even if this does not solve the lifetime issue you mentioned I'm still having a headache, why **inline** or **not inline** is the differentiator between working and not working code. I would expect a lifetime issue to be recognizable regardless of the function being inlined or not ?

**Comment by gnprice** [comment]
> why **inline** or **not inline** is the differentiator between working and not working code.

Once you're in the realm of undefined behavior, your program may happen to work and may not. Inlining or un-inlining a function are among the many things that could cause a bug to hide or re-emerge.

For help debugging your program, I'd encourage you to try posting on https://users.rust-lang.org/ .
